### ISSUE TITLE
Unit Test Framework

### ISSUE THREAD

**Issue by thomaskeller79** [prost-planner/prost#62]
**[Original report](https://bitbucket.org/tkeller/prost/issue/62) by geisserf (Bitbucket: [557058:e7a9f9a5-3ea8-4154-97d2-10446425dce3](https://bitbucket.org/557058:e7a9f9a5-3ea8-4154-97d2-10446425dce3), GitHub: [geisserf](https://github.com/geisserf)).**

----------------------------------------

Before I start implementing fixes for the IPPC18 issues I would like to re-initiate the discussion about unit tests (see discussion in issue #38), to agree on a unit test framework.

Previously, we used the [Google test framework gtest](https://github.com/google/googletest). I've worked with the [Catch library](https://github.com/catchorg/Catch2) on other projects, which also works quite well. Since Catch is a header-only library, it is a little bit simpler to include as a new user (i.e. when checking out the repository), and probably also easier to support under Windows (this is just a hunch, though). Feature and usability wise, both libraries seem to perform quite well.

**Comment by geisserf** [comment]
When I was writing a unit test for a parser component I stumbled on something which I think might be a potential bug. I pushed the respective test to the current branch. In src/rddl_parser/tests/probability_distribution_test.cc is a test for the functionality of the probability distribution class. If you build, the test in line 35 currently fails. The reason is that the discrete distribution assigns two probabilities to the same value, which is overwritten in the passed map.

Clearly, instead of assigning the same value in a distribution two times it would be sufficient to sum up the probabilities and only pass the value a single time. But in principle there is nothing wrong with designing a distribution that is like { 1 : 0.2, 1 : 0.2, 2 : 0.6 }, would you agree?

**Comment by thomaskeller79** [comment]
In principle, nothing is wrong with { 1 : 0.2, 1 : 0.2, 2 : 0.6 }, but then we'd have to put additional effort into the usage of the class. For instance, one chance node is created for each entry in the PD, and if two values are equal, THTS would create two successors that contain the same state. While this is not wrong, it's not what we want.
I believe there is already some code that combines probabilities if they map to the same value. It would make sense to change the DiscretePD implementation such that probabilities that are assigned to the same value are sumed up *inside* the class rather than to expect this is done outside.

**Comment by geisserf** [comment]
I guess we should make a separate issue for this.

I moved the doctest header file into src/ instead src/rddl_parser/tests and src/search/tests, because otherwise it would be duplicated.

**Comment by geisserf** [comment]
I know we talked about having the doctest header in both tests/ directories, but now that I had to add the license file I created a separate src/doctest directory which includes doctest as well as its license. This seems to be sufficient for the license, and once we decide on which license the planner itself uses we can add a note similar to how Fast Downward does it that says "the following directory is not part of Prost as covered by this license".

From my side this issue is then finished. Should I make a new branch on the git prost-dev repo, add the changes I have in the hg prost-dev repo, and then open a pull request?
